### PR TITLE
Damn it all to Hell! Adds jinxing and blessing for the Chaplain, loose framework for more dice rolls

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
Adds a new component, rngod, that allows the user to curse or bless nearby mobs. Curses and blessings give modifiers to certain dice rolls that someone may come across, either helping or hurting their outcomes! It currently works by saying either "damn" (for curses) or "bless" (for blessings), optionally along with the name of someone specific to target. The component checks the surrounding area for your target, and if it can't find them, it applies the modifier to a random nearby person (including possibly yourself, in the case of curses). The modifier is then entered into the target's "fate" queue (a new datum stored in the mind), which is checked every time they make a qualifying roll. 

This ability is available only to the Chaplain at roundstart, but I think there's tons of room for expanding into other items/traits/whatever that would also be able to grant modifiers. If you think someone's stacked a bunch of curses on you, you can clear your fate queue by drinking holy water.

These modifiers assume the rolls we're dealing with follow a certain format that I've loosely used in a few of my own PRs. The rolls should be percentile (based on 1-100), and use degrees of success (different thresholds for how well you roll, the higher you roll the better the outcome, the lower you roll the worse). The only roll that does this in game right now is the one for attacking vendors, where better rolls get you more free stuff and worse rolls get you maimed, but my tackling PR (#49476) follows this format as well for rolling injuries when hitting obstacles (also probably the mob vs mob roll-off roll, though that's not percentile right now). I've had a lot of fun designing these systems, so I think they could be fun put in other places as well.

## Why It's Good For The Game
Adds more ways to help/hinder your fellow crewmembers, intentionally and unintentionally. "Damn" was chosen as the curse triggerword specifically because it's easy to activate carelessly (as well as being thematically relevant), creating a new source of social faux-pas' and misunderstandings for Chaplains and whoever else might have this. This also gives the Chaplain a source of soft social power, since slighting them can come back to bite you later without you even realizing it.

Right now this PR is pretty toothless and limited in scope, but I wanted to solicit feedback on if people think it's worth expanding on, because I have lots of ideas. Some of my favorite parts of tabletop roleplaying games are the tables for roll results, especially when it comes to rare crit success/failure outcomes that do really exciting/unexpected things. Lots of people in the tgstation community are bearish about making roundstart job roles/skills too integral to being able to do things effectively, so I think that adding modifiers to things like equipment, traits, supernatural connections, and things of that nature is the way to go with adding RNG in a fun and manageable way. Really, this would be the first step in adding a more complete framework for dice rolls and modifiers that I think would make a really positive impact on the game, so I'm treating this PR as a referendum of sorts on the matter (obviously dice rolls wouldn't be jammed into everything carelessly, I envision them being used to determine how efficiently/effectively someone does something, not a hard line on whether they can do it or not, barring extremely bad rolls failing spectacularly).

And of course, the implementation specifics are all subject to change and should be treated as a general draft for the time being.

## Changelog
:cl: Ryll/Shaps
todo when finalized
/:cl:

### CODE REVIEW

File: code/datums/components/rngod.dm [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+	for(var/mob/M in view(radius, parent)) // look through all the mobs nearby
+		var/list/nameWords = list()
+		if(!M.mind || !M.mind.fate)
+			continue
+
+		for(var/string in splittext(lowertext(M.real_name), " "))
+			nameWords += string
+		for(var/string in splittext(lowertext(M.name), " "))
+			nameWords += string
```
Comment (XDTM) [inline code/datums/components/rngod.dm]: i made a first_name() helper for voice of god, you could check it here so you can curse someone without having to get their full name right

Comment (trollbreeder) [reply code/datums/components/rngod.dm]: `M.first_name()`

Comment (Ryll-Ryll) [reply code/datums/components/rngod.dm]: It splits on spaces anyway, so getting any of the individual parts of their name works

Comment (XDTM) [reply code/datums/components/rngod.dm]: ah, i didn't even notice the splittext, my bad

it doesn't catch lizard names, but you could fix that by replacing dashes with spaces beforehand

**Comment by ATH1909** [comment]
Since the sleeping carp rework PR got merged, maybe you could make this affect the crit chance of sleeping carp strong punches too?

**Comment by ATH1909** [comment]
Ooh, and there's quite a bit of attack_alien() RNG in the game too, spread across various types of mobs.

**Comment by Qustinnus** [comment]
I'd prefer this to be rolled into sects rather than it being its own thing

**Comment by necromanceranne** [comment]
Oh hecc it's Bless and Bane for chappies. Nice.

**Comment by Spepbuthewranglesgithub** [comment]
do it, curse the clown with _double clumsiness,_ i know you want to.

**Comment by MrDoomBringer** [comment]
The chaplain dying should break any curses/blessings

File: code/datums/components/rngod.dm (hunk 1 above)
Comment (spessbandit) [reply code/datums/components/rngod.dm]: some lizards don't go by their first names though (c.f. Is-a-Lizard)

Comment (spessbandit) [reply code/datums/components/rngod.dm]: what happens when people share individual parts of names, like, Eats-the-Lemons and Smells-the-Feet

**Comment by ATH1909** [comment]
> do it, curse the clown with _double clumsiness,_ i know you want to.

What if we made a curse-focused chaplain sect later that could curse people with (temporary) clumsiness?

File: code/datums/components/rngod.dm (hunk 1 above)
Comment (Ryll-Ryll) [reply code/datums/components/rngod.dm]: These are the RNGods, they just stochastically pick one! Well, whoever comes up first in the search anyway

**Comment by spessbandit** [comment]
hell yes we apocalypse world now

**Comment by spessbandit** [comment]
this would be good for revenant

File: code/datums/components/rngod.dm (hunk 1 above)
Comment (spessbandit) [reply code/datums/components/rngod.dm]: is viewers pre-randomized? (can people powergame their names to avoid this)

**Comment by Couls** [comment]
Thanking should also bless because people don’t use the word bless as often

File: code/datums/components/rngod.dm (hunk 1 above)
Comment (XDTM) [reply code/datums/components/rngod.dm]: imagine powergaming your name so the chaplain can't make it slightly more likely that a vending machine squishes you

plus, if you make it more likely to get cursed you also make it less likely to get blessed and viceversa

**Comment by XDTM** [comment]
After this is merged i'm gonna add a divination mechanic where drawing a card from a tarot deck or looking at a crystal ball prepares a fate modifier

**Comment by Ty-the-Smonk** [comment]
Scientists are all atheists so they are immune

**Comment by Farquaar** [comment]
Not really a fan of this.

**Comment by Ryll-Ryll** [comment]
Considering sects are still heavily WIP and the creator said they want to finish the framework first, as well as the fact that this is almost entirely a passive mechanic (more like a trait) and sects are very much an active mechanic, I see no reason to make it a sect at this time. Having reread the sects PR it sounds like cobby has something totally different in mind than how this works, so other than them both concerning the chaplain I don't see any connection. If later on someone wants to make a sect that has a more powerful or flexible version of this, they can do so

I'll look into the other suggestions about name handling, thanks instead of bless, and more rolls in general

File: code/datums/components/rngod.dm (hunk 1 above)
Comment (ATH1909) [reply code/datums/components/rngod.dm]: You should still shuffle it if it's just based on alphabetical order, though.

File: code/datums/components/rngod.dm [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  *
+  * Arguments:
+  * *msg- The lowercase'd message string we're digging through for names
+  * *radius- How far we're looking
+  */
+/datum/component/rngod/proc/findTarget(msg, radius=3)
+	//explode the input msg into a list
+	var/list/msglist = splittext(msg, " ")
+
+	for(var/mob/M in view(radius, parent)) // look through all the mobs nearby
```
Comment (LudwigVonChesterfield) [inline code/datums/components/rngod.dm]: wouldn't for(var/mob/M in viewers()) be more correct?

File: code/datums/components/fate.dm [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+/datum/component/fate
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+
+	/// when did we last bless someone?
+	var/list/queue = list()
+
+
+/datum/component/fate/Initialize()
+
+/datum/component/fate/Destroy(force, silent)
```
Comment (LudwigVonChesterfield) [inline code/datums/components/fate.dm]: why would you override a proc, if the only thing you're doing is returning what parent returns?

File: code/datums/components/fate.dm [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  * *
+  */
+/datum/component/fate
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+
+	/// when did we last bless someone?
+	var/list/queue = list()
+
+
+/datum/component/fate/Initialize()
```
Comment (LudwigVonChesterfield) [inline code/datums/components/fate.dm]: this override is pointless, since proc's body contains nothing

File: code/datums/components/fate.dm [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	return roll
+
+/**
+  * Here you can add a modifier to someone's fate queue, which will give them a modifier to whatever roll this ends up being called for
+  *
+  * Arguments:
+  * * datum/source: What was the source?
+  * * amount: How much will we add to the roll this is called for?
+  */
+/datum/component/fate/proc/addModifier(datum/us, amount)
```
Comment (LudwigVonChesterfield) [inline code/datums/components/fate.dm]: would naming this adjustModifier be more appropriate, since amount can be negative?

**Comment by PKPenguin321** [comment]
What ever happened to us using seeded RNG? We should get that going again and roll modifiers like this into it

**Comment by ExcessiveUseOfCobblestone** [comment]
Sects are ready to go into prod it just doesn't have a variety of sects since I wanted to focus on adding framework first (and I don't plan on adding more in that PR).

I can see alignment as well as the keywords changing per the sect, IE if you were the technophile sect you might curse someone by saying you hope they never be made a borg or something.

**Comment by Ryll-Ryll** [comment]
Honestly after looking up the seeded rng PR pkp mentioned (#30087), I feel inspired to reach for the stars and expand into making an RNG subsystem. I'm gonna mull on this for a bit

**Comment by optimumtact** [comment]
I don't really see much of a place for RNG rolls, your tackle thing already crosses lines on stun related combat with RNG rolling.

If you do want to actually make what is effectively a luck or karma system, you need some mechanism for feeding back to the user how lucky they are right now.

**Comment by necromanceranne** [comment]
Would a moodlet work or is that too subtle?

**Comment by PKPenguin321** [comment]
Make it a skill

**Comment by Ryll-Ryll** [comment]
This is premature, definitely a lot to figure out before it's near workable

**Comment by optimumtact** [comment]
make it a big scale that tilts as you get bad or good karma

**Pull request closed without merging**